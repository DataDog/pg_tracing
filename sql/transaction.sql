-- Only trace queries with sample flag
SET pg_tracing.sample_rate = 0.0;
SET pg_tracing.caller_sample_rate = 1.0;

-- Set tracecontext at the start of the transaction
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000001-0000000000000001-01'*/ BEGIN; SELECT 1; COMMIT;

SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000001' AND span_type!='Commit';
CALL clean_spans();

-- Test with override of the trace context in the middle of a transaction
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000002-0000000000000002-01'*/ BEGIN;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000003-0000000000000003-01'*/ SELECT 1; COMMIT;

SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000002' AND span_type!='Commit';
SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000003';
CALL clean_spans();

-- Test with a 0 traceid in the middle of a transaction
BEGIN;
SELECT 1;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000001-01'*/ SELECT 2;
SELECT 3;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000001-01'*/ SELECT 4;
END;

-- Only one trace id should have been generated
SELECT count(distinct(trace_id)) = 1 FROM pg_tracing_peek_spans;
SELECT span_type, span_operation, lvl FROM peek_ordered_spans;
CALL clean_spans();

-- Test with a 0 parent_id in the middle of a transaction
BEGIN;
SELECT 1;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000000-01'*/ SELECT 2;
SELECT 3;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000000-01'*/ SELECT 4;
END;

-- Only one trace id and parent id should have been generated
SELECT count(distinct(trace_id)) = 1, count(distinct(parent_id)) = 1 FROM peek_ordered_spans WHERE lvl=1;
SELECT span_type, span_operation, lvl FROM peek_ordered_spans;

CALL clean_spans();
CALL reset_settings();
