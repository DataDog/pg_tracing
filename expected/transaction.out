-- Only trace queries with sample flag
SET pg_tracing.sample_rate = 0.0;
SET pg_tracing.caller_sample_rate = 1.0;
-- Set tracecontext at the start of the transaction
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000001-0000000000000001-01'*/ BEGIN; SELECT 1; COMMIT;
 ?column? 
----------
        1
(1 row)

SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000001' AND span_type!='Commit';
   span_type    | span_operation | lvl 
----------------+----------------+-----
 Utility query  | BEGIN;         |   1
 ProcessUtility | ProcessUtility |   2
 Select query   | SELECT $1;     |   1
 Planner        | Planner        |   2
 ExecutorRun    | ExecutorRun    |   2
 Result         | Result         |   3
 Utility query  | COMMIT;        |   1
 ProcessUtility | ProcessUtility |   2
(8 rows)

CALL clean_spans();
-- Test with override of the trace context in the middle of a transaction
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000002-0000000000000002-01'*/ BEGIN;
/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000003-0000000000000003-01'*/ SELECT 1; COMMIT;
 ?column? 
----------
        1
(1 row)

SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000002' AND span_type!='Commit';
   span_type    | span_operation | lvl 
----------------+----------------+-----
 Utility query  | BEGIN;         |   1
 ProcessUtility | ProcessUtility |   2
 Utility query  | COMMIT;        |   1
 ProcessUtility | ProcessUtility |   2
(4 rows)

SELECT span_type, span_operation, lvl FROM peek_ordered_spans where trace_id='00000000000000000000000000000003';
  span_type   | span_operation | lvl 
--------------+----------------+-----
 Select query | SELECT $1;     |   1
 Planner      | Planner        |   2
 ExecutorRun  | ExecutorRun    |   2
 Result       | Result         |   3
(4 rows)

CALL clean_spans();
-- Test with a 0 traceid in the middle of a transaction
BEGIN;
SELECT 1;
 ?column? 
----------
        1
(1 row)

/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000001-01'*/ SELECT 2;
 ?column? 
----------
        2
(1 row)

SELECT 3;
 ?column? 
----------
        3
(1 row)

/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000001-01'*/ SELECT 4;
 ?column? 
----------
        4
(1 row)

END;
-- Only one trace id should have been generated
SELECT count(distinct(trace_id)) = 1 FROM pg_tracing_peek_spans;
 ?column? 
----------
 t
(1 row)

SELECT span_type, span_operation, lvl FROM peek_ordered_spans;
  span_type   | span_operation | lvl 
--------------+----------------+-----
 Select query | SELECT $1;     |   1
 Planner      | Planner        |   2
 ExecutorRun  | ExecutorRun    |   2
 Result       | Result         |   3
 Select query | SELECT $1;     |   1
 Planner      | Planner        |   2
 ExecutorRun  | ExecutorRun    |   2
 Result       | Result         |   3
(8 rows)

CALL clean_spans();
-- Test with a 0 parent_id in the middle of a transaction
BEGIN;
SELECT 1;
 ?column? 
----------
        1
(1 row)

/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000000-01'*/ SELECT 2;
 ?column? 
----------
        2
(1 row)

SELECT 3;
 ?column? 
----------
        3
(1 row)

/*dddbs='postgres.db',traceparent='00-00000000000000000000000000000000-0000000000000000-01'*/ SELECT 4;
 ?column? 
----------
        4
(1 row)

END;
-- Only one trace id and parent id should have been generated
SELECT count(distinct(trace_id)) = 1, count(distinct(parent_id)) = 1 FROM peek_ordered_spans WHERE lvl=1;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

SELECT span_type, span_operation, lvl FROM peek_ordered_spans;
  span_type   | span_operation | lvl 
--------------+----------------+-----
 Select query | SELECT $1;     |   1
 Planner      | Planner        |   2
 ExecutorRun  | ExecutorRun    |   2
 Result       | Result         |   3
 Select query | SELECT $1;     |   1
 Planner      | Planner        |   2
 ExecutorRun  | ExecutorRun    |   2
 Result       | Result         |   3
(8 rows)

CALL clean_spans();
CALL reset_settings();
